R/data-combined.R

```diff
@@ -346,6 +349,144 @@ DataCombined <- R6::R6Class(
+    #' selected. In such cases, only the first of these is retained. For
+    #' example, with `threshold = 1`, two simulated values (`c(1.4, 2.2`) are
+    #' selected and only the first one (`1.4`) is retained.
+    #'
+    #' On the other hand, if the threshold is too **stringent**, no simulated
+    #' values might be selected. With a `threshold = 0.1`, no simulated values
+    #' will be selected for out example data.
+    #'
+    #' @param threshold Absolute numerical distance by which the closest
+    #'   `xValue` (time, i.e.) in the observed data may differ from `xValue` in
+    #'   simulated data to be accepted as *equal*. If no threshold is specified
+    #'   (default), then the square root of smallest positive floating-point
+    #'   number on your machine will be used (`.Machine$double.eps^0.5`) as a
+    #'   threshold. For all intents and purposes, this amounts to an **exact**
+    #'   match between two values being compared.
+    toObsVsPredDataFrame = function(threshold = NULL) {
+      validateIsNumeric(threshold, nullAllowed = TRUE)
+
+      # We don't want this function to tamper with combined dataframe, so
+      # make an internal copy for this function scope.
+      df <- private$.dataCombined %>%
+        # Remove unnecessary columns
+        private$.cleanDF(.) %>%
+        # Datasets which are not part of any grouping can't be paired
+        dplyr::filter(!is.na(group))
+
+      if (is.null(df)) {
+        stop(
+          "There are currently no datasets. You can add them with `$addDataSets()` and/or `$addSimulationResults()` methods.",
+          call. = FALSE
+        )
+      }
+
+      if (length(unique(df$group)) == 0L) {
+        stop(
+          "There are currently no grouping specified. You can set them with `$setGroups()` method.",
+          call. = FALSE
+        )
+      }
+
+      if (length(unique(df$dataType)) == 1L && unique(df$dataType) == "simulated") {
+        stop(
+          "Both observed and simulated data needed, but only simulated data is present.",
+          call. = FALSE
+        )
+      }
+
+      if (length(unique(df$dataType)) == 1L && unique(df$dataType) == "observed") {
+        stop(
+          "Both observed and simulated data needed, but only observed data is present.",
+          call. = FALSE
+        )
+      }
+
+      # Unlike the dataframe returned by `$toDataFrame()` method, this dataframe
+      # needs to be in wide - and not tidy - format because the `xValues` for
+      # observed versus simulated datasets need to be paired for plotting.
+      #
+      # Additionally, the used plotting library (`{tlf}`) also expects a data in
+      # the wide format.
+      #
+      # This entails that if both of these wide dataframes need to be joined at
+      # some point, they should also have unique names.
+      obsData <- df %>%
+        dplyr::filter(dataType == "observed") %>%
+        dplyr::rename_with(~ paste0(.x, "Observed"), -dplyr::matches("^group$"))
```

Comment:
I don't understand what is it you are doing here? You are adding Observed and Predicted to the column name to avoid duplication? 
Also why not Simulated (vs Predicted since the dataType is Simulated)

Reply:
> You are adding Observed and Predicted to the column name to avoid duplication?

It's to highlight which type of data the values are associated with. For example, both for observed and simulated data, the time values are in `xValues` column, and we need to distinguish them in a wide format dataframe by having two columns `xValuesObserved` and `xValuesPredicted`.

Here is a simple example:

``` r
library(tidyverse, warn.conflicts = FALSE)

# tidy ---------

df1 <- tibble(name = "a", group = "x")
df2 <- tibble(name = "b", group = "y")

dplyr::full_join(df1, df2, by = "group")
#> # A tibble: 2 x 3
#>   name.x group name.y
#>   <chr>  <chr> <chr> 
#> 1 a      x     <NA>  
#> 2 <NA>   y     b

# wide ---------

df1 <- dplyr::rename_with(df1, ~ paste0(.x, "Observed"), -dplyr::matches("^group$"))
df2 <- dplyr::rename_with(df2, ~ paste0(.x, "Predicted"), -dplyr::matches("^group$"))

dplyr::full_join(df1, df2, by = "group")
#> # A tibble: 2 x 3
#>   nameObserved group namePredicted
#>   <chr>        <chr> <chr>        
#> 1 a            x     <NA>         
#> 2 <NA>         y     b
```

<sup>Created on 2022-03-07 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

We need data to be in the wide format because that's the format _tlf_ function expects.

> Also why not Simulated (vs Predicted since the dataType is Simulated)

No strong opinions either way. I can change it to `Simulated` if so desired.

I chose Predicted only because the plotting function is going to be called `plotPredictedVersusObserved()`, and not `plotSimulatedVersusObserved()`. So thought better to be consistent with naming.

But I am totally flexible here.

Reply:
> We need data to be in the wide format because that's the format tlf function expects.

I was initially against having this method public and would prefer that such a DF is created internally by the `plotXXXVsObserved`, but you argumented that the user might want to have this information independently of calling `plotXXXVsObserved` because he might want to create his own plot... But now you are tailoring the output exactly to the use with TLF within the function to be created, and not for the general purpose. And now we end up in two completely different output formats (`$toDataFrame` and `$toObsVsPredDataFrame`) but containing actually the same information. I would be (and am) very confused.

Reply:
> But now you are tailoring the output exactly to the use with TLF within the function to be created, and not for the general purpose.

Nope. You _need_ the data to be in the wide format _irrespective of_ what package you choose, doesn't matter if it's _tlf_, _ggplot2_, or _base_. Paired data that needs to be plotted as a scatterplot will always have to be in wide format. It will be confusing for it _not_ to be.

I am mentioning _tlf_ here only because that's the one we will be using. The story is going to be the same no matter what you use.

```r
# ggplot2

ggplot(df, aes(xValuesObserved, xValuesPredicted)) + 
  geom_point()

# base

plot(df$xValuesObserved, df$xValuesPredicted)
```

Reply:
cool thanks